File: src/main/java/corp/soyummy/shoppinglists/ShoppingListController.java
package corp.soyummy.shoppinglists;

import org.springframework.web.bind.annotation.*;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import lombok.RequiredArgsConstructor;
import jakarta.validation.Valid;
import java.util.List;

import corp.soyummy.auth.User;
import corp.soyummy.shoppinglists.dto.ShoppingListCreateDto;
import corp.soyummy.shoppinglists.dto.ShoppingListUpdateDto;
import corp.soyummy.errors.ResourceNotFoundException;
import static corp.soyummy.constants.Constants.VERSION;

@RestController
@RequiredArgsConstructor
@RequestMapping(VERSION + "shoppinglists")
public class ShoppingListController {

    private final ShoppingListService shoppingListService;

    //  GET  http://localhost:8080/api/v1/shoppinglists
    @GetMapping
    public ResponseEntity<List<ShoppingList>> getAllShoppingLists(@AuthenticationPrincipal User user) {
        List<ShoppingList> shoppingLists = shoppingListService.getAllShoppingLists();
        return ResponseEntity.ok(shoppingLists);
    }

    //  GET  http://localhost:8080/api/v1/shoppinglists/{id}
    @GetMapping("/{id}")
    public ResponseEntity<ShoppingList> getShoppingListById(
            @PathVariable String id,
            @AuthenticationPrincipal User currentUser) {
        ShoppingList shoppingList = shoppingListService.getShoppingListById(id);
        if (!shoppingList.getOwner().equals(currentUser.getId())) {
            throw new ResourceNotFoundException("Shopping list not found for current user");
        }

        return ResponseEntity.ok(shoppingList);
    }

    //  GET  http://localhost:8080/api/v1/shoppinglists/byOwner/{ownerId}
    @GetMapping("/byOwner/{ownerId}")
    public ResponseEntity<List<ShoppingList>> getShoppingListsByOwner(
            @PathVariable String ownerId,
            @AuthenticationPrincipal User currentUser) {
        if (!ownerId.equals(currentUser.getId())) {
            throw new ResourceNotFoundException("You can only access your own shopping list");
        }

        List<ShoppingList> shoppingLists = shoppingListService.findByOwner(ownerId);
        return ResponseEntity.ok(shoppingLists);
    }

    //  GET  http://localhost:8080/api/v1/shoppinglists/byRecipeId/{recipeId}
    @GetMapping("/byRecipeId/{recipeId}")
    public ResponseEntity<List<ShoppingList>> getShoppingListsByRecipeId(@PathVariable String recipeId) {
        List<ShoppingList> shoppingLists = shoppingListService.findByRecipeId(recipeId);
        return ResponseEntity.ok(shoppingLists);
    }

    //  GET  http://localhost:8080/api/v1/shoppinglists/byStartIngredient/{startIngredient}
    @GetMapping("/byStartIngredient/{startIngredient}")
    public ResponseEntity<List<ShoppingList>> getShoppingListsByStartIngredient(@PathVariable String startIngredient) {
        List<ShoppingList> shoppingLists = shoppingListService.findByStartIngredient(startIngredient);
        return ResponseEntity.ok(shoppingLists);
    }

    //  POST  http://localhost:8080/api/v1/shoppinglists
    @PostMapping
    public ResponseEntity<ShoppingList> createShoppingList(
            @Valid @RequestBody ShoppingListCreateDto shoppingListCreateDto,
            @AuthenticationPrincipal User currentUser) {
        shoppingListCreateDto.setOwner(currentUser.getId());
        ShoppingList createdShoppingList = shoppingListService.createShoppingList(shoppingListCreateDto);

        return ResponseEntity.status(201).body(createdShoppingList);
    }

    //  PUT  http://localhost:8080/api/v1/shoppinglists/{id}
    @PutMapping("/{id}")
    public ResponseEntity<ShoppingList> updateShoppingList(
            @PathVariable String id,
            @Valid @RequestBody ShoppingListUpdateDto shoppingListUpdateDto,
            @AuthenticationPrincipal User currentUser) {
        ShoppingList existingShoppingList = shoppingListService.getShoppingListById(id);
        if (!existingShoppingList.getOwner().equals(currentUser.getId())) {
            throw new ResourceNotFoundException("Shopping list not found for current user");
        }

        ShoppingList updatedShoppingList = shoppingListService.updateShoppingList(id, shoppingListUpdateDto);
        return ResponseEntity.ok(updatedShoppingList);
    }

    //  DELETE  http://localhost:8080/api/v1/shoppinglists/{id}
    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteShoppingList(
            @PathVariable String id,
            @AuthenticationPrincipal User currentUser) {
        ShoppingList existingShoppingList = shoppingListService.getShoppingListById(id);
        if (!existingShoppingList.getOwner().equals(currentUser.getId())) {
            throw new ResourceNotFoundException("Shopping list not found for current user");
        }

        shoppingListService.deleteShoppingList(id);
        return ResponseEntity.noContent().build();
    }
}

File: src/main/kotlin/com/soyummy/shoppinglists/ShoppingListController.kt
package com.soyummy.shoppinglists

import org.springframework.web.bind.annotation.*
import org.springframework.security.access.annotation.Secured
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.http.ResponseEntity
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import jakarta.validation.Valid

import com.soyummy.common.dto.PageResponse
import com.soyummy.shoppinglists.dto.ShoppingListCreateDto
import com.soyummy.shoppinglists.dto.ShoppingListUpdateDto
import com.soyummy.auth.User
import com.soyummy.constants.Constants.VERSION

@RestController
@RequestMapping("${VERSION}/shoppinglists")
class ShoppingListController(private val shoppingListService: ShoppingListService) {

    //  GET  http://localhost:8080/api/v1/shoppinglists?page=3&size=4
    @Secured("ROLE_ADMIN")
    @GetMapping
    fun getAllShoppingLists(
        @RequestParam(defaultValue = "0") page: Int,
        @RequestParam(defaultValue = "10") size: Int,
        @RequestParam(defaultValue = "owner,asc") sort: String
    ): ResponseEntity<PageResponse<ShoppingList>> {
        val sortParams = sort.split(",")
        val sortDirection = if (sortParams.size > 1 && sortParams[1].lowercase() == "desc") {
            Sort.Direction.DESC
        } else {
            Sort.Direction.ASC
        }
        val sortField = sortParams[0]

        val pageable = PageRequest.of(page, size, Sort.by(sortDirection, sortField))
        val result = shoppingListService.getAllShoppingLists(pageable)

        return ResponseEntity.ok(result)
    }

    //  GET  http://localhost:8080/api/v1/shoppinglists/byOwner/{ownerId}
    @GetMapping("/byOwner/{ownerId}")
    fun getShoppingListsByOwnerId(
        @PathVariable ownerId: String,
        @AuthenticationPrincipal currentUser: User?
    ): ResponseEntity<List<ShoppingList>> =
        ResponseEntity.ok(shoppingListService.findByOwner(ownerId, currentUser))

    //  GET  http://localhost:8080/api/v1/shoppinglists/byRecipeId/{recipeId}
    @GetMapping("/byRecipeId/{recipeId}")
    fun getShoppingListsByRecipeId(
        @PathVariable recipeId: String,
        @AuthenticationPrincipal currentUser: User?
    ): ResponseEntity<List<ShoppingList>> =
        ResponseEntity.ok(shoppingListService.findByRecipeId(recipeId, currentUser))

    //  GET  http://localhost:8080/api/v1/shoppinglists/byStartIngredient/{startIngredient}
    @GetMapping("/byStartIngredient/{startIngredient}")
    fun getShoppingListsByStartIngredient(
        @PathVariable startIngredient: String,
        @AuthenticationPrincipal currentUser: User?
    ): ResponseEntity<List<ShoppingList>> =
        ResponseEntity.ok(shoppingListService.findByStartIngredient(startIngredient, currentUser))

    //  POST  http://localhost:8080/api/v1/shoppinglists
    @PostMapping
    fun createShoppingList(
        @Valid @RequestBody shoppingListCreateDto: ShoppingListCreateDto,
        @AuthenticationPrincipal currentUser: User?
    ): ResponseEntity<ShoppingList> =
        ResponseEntity.status(201).body(shoppingListService.createShoppingList(shoppingListCreateDto, currentUser))

    //  PUT  http://localhost:8080/api/v1/shoppinglists/{id}
    @PutMapping("/{id}")
    fun updateShoppingList(
        @PathVariable id: String,
        @Valid @RequestBody shoppingListUpdateDto: ShoppingListUpdateDto,
        @AuthenticationPrincipal currentUser: User?
    ): ResponseEntity<ShoppingList> =
        ResponseEntity.ok(shoppingListService.updateShoppingList(id, shoppingListUpdateDto, currentUser))

    //  DELETE  http://localhost:8080/api/v1/shoppinglists/{id}
    @DeleteMapping("/{id}")
    fun deleteShoppingList(
        @PathVariable id: String,
        @AuthenticationPrincipal currentUser: User?
    ): ResponseEntity<Void> {
        shoppingListService.deleteShoppingList(id, currentUser)
        return ResponseEntity.noContent().build()
    }
}